components/ee/ws-scheduler/pkg/scheduler/scheduler.go

```diff
@@ -623,57 +715,120 @@ func queuePodForScheduling(schedulerQueue chan<- *corev1.Pod, pod *corev1.Pod) {
 	}
 }
 
-// localBindingCache stores whether we already scheduled a Pod and is necessary to bridge the gap between:
+func isKubernetesObjNotFoundError(err error) bool {
+	if err, ok := err.(*k8serr.StatusError); ok {
+		return err.ErrStatus.Code == http.StatusNotFound
+	}
+	return false
+}
+
+// localSlotCache stores whether we already scheduled a Pod and is necessary to bridge the gap between:
 //  1. We successfully create a binding with the Kubernets API
 //  2. The change is reflected in the results we get from s.pods.Lister().List()
 // Without it, we sometimes end up with pods being assigned the same last slot on a node, leading
 // to workspace startup failures.
 //
 // (We're not sure why there is gap in the first place: Kubernetes should be consistent, as etcd is
 // consistent (per Object type, here Pod). Maybe because of the GCloud multi-master setup..?)
-type localBindingCache struct {
-	bindings map[string]*Binding
-	mu       sync.RWMutex
+type localSlotCache struct {
+	slots map[string]slot
+	mu    sync.RWMutex
+}
+
+// slot is used to describe a "place" on a single node that we already scheduled/are trying to schedule a workspace on.
+type slot struct {
+	binding       *Binding
+	reservedGhost string
 }
 
-func (c *localBindingCache) markAsScheduled(pod *corev1.Pod, nodeName string) {
+// markAsScheduled marks a certain slot as already scheduled
+func (c *localSlotCache) markAsScheduled(pod *corev1.Pod, nodeName string, ghostToReplace string) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	c.bindings[pod.Name] = &Binding{
-		Pod:      pod,
-		NodeName: nodeName,
+	c.slots[pod.Name] = slot{
+		binding: &Binding{
+			Pod:      pod,
+			NodeName: nodeName,
+		},
+		reservedGhost: ghostToReplace,
 	}
 }
 
-func (c *localBindingCache) hasAlreadyBeenScheduled(podName string) bool {
+func (c *localSlotCache) hasAlreadyBeenScheduled(podName string) bool {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
-	_, present := c.bindings[podName]
+	_, present := c.slots[podName]
 	return present
 }
 
-func (c *localBindingCache) delete(podName string) {
+func (c *localSlotCache) delete(podName string) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	delete(c.bindings, podName)
+	delete(c.slots, podName)
 }
 
-func (c *localBindingCache) getListOfBindings() []*Binding {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (c *localSlotCache) getListOfBindings() []*Binding {
+	c.mu.RLock()
```

Comment:
@csweichel I changed the lock from "write" to "read" here. I'm unsure why it was write in the first place... any idea? :thinking: